website/integrations: add documentation for OIDC setup with Xen Orchestra (#9000)
goauthentik/authentik: website/integrations/services/xen-orchestra/index.md

---
title: Xen Orchestra
---

<span class="badge badge--secondary">Support level: Community</span>

## What is Xen Orchestra

> Xen Orchestra provides a user friendly web interface for every Xen based hypervisor (XenServer, xcp-ng, etc.).
>
> -- https://xen-orchestra.com/

:::note
Xen Orchestra offers authentication plugins for OpenID Connect, SAML and LDAP. This guide is usind the OpenID Connect plugin.
If you are using the Xen Orchestra Appliance, the OIDC Plugin should be present. If you are using Xen Orchestra compiled from sources, make sure the plugin `auth-oidc` is installed.
:::

## Preparation

The following placeholders will be used:

-   `xenorchestra.company` is the FQDN of the Xen Orchestra instance.
-   `authentik.company` is the FQDN of the authentik install.

## authentik configuration

### 1. Provider

Under _Providers_, create an OAuth2/OpenID provider with these settings:

-   Name: Provider for XenOrchestra
-   Authorization Flow: Select one of the availible Flows.
-   Client type: Confidental
-   Redirect URIs/Origins: `https://xenorchestra.company/signin/oidc/callback`

> Take note of the Client ID and the Client Secret, because we need them for the Configuration in Xen Orchestra.

Reviewer: ```suggestion
Take note of the Client ID and the Client Secret, because we need them for the configuration of Xen Orchestra.
```

### 2. Application

Create an application with the following details:

> -   Slug: `xenorchestra` (If you want to choose a different slug, your URLs for the Xen Orchestra Configuration may vary)

Reviewer: ```suggestion
-   Slug: `xenorchestra` (If you want to choose a different slug, your URLs for the Xen Orchestra Configuration may vary.)
```

-   Provider: Select the one we have created in Step 1

> -   Set the Launch URL to `https://xenorchestra.company/`.

Reviewer: ```suggestion
-   Set the Launch URL to `https://xenorchestra.company/`
```

Optionally apply access restrictions to the application.

## Xen Orchestra configuration

Xen Orchestra allows the configuration of the OpenID Connect authentication in the plugin-section.

> All of the URLs metioned below can be copied & pasted from authentik (Applications -> Providers -> the provider created earlier)

Reviewer: ```suggestion
All of the URLs mentioned below can be copied & pasted from authentik (_Applications -> Providers -> *the provider created earlier*_).
```

1. Navigate to Settings -> Plugins

> 2. Scroll to "auth-oidc" and click on the "+ Icon" on the right hand side.

Reviewer: ```suggestion
2. Scroll to **auth-oidc** and click on the **+** icon on the right hand side.
```

> 3. Set the `Auto-discovery URL` to `https://authentik.company/application/o/xenorchestra/.well-known/openid-configuration`

Reviewer: ```suggestion
3. Set the `Auto-discovery URL` to `https://authentik.company/application/o/xenorchestra/.well-known/openid-configuration`.
```

Reviewer: Up to you, but where possible, you can simplify the docs by grouping the stps all into a single step, if they are done on the same UI page in Xen Orchestra. (Sorry I don't have access to the XO Ui or I would make a more exact suggestion.) Basically, if a group of settings are on the same UI, you can format it like this:

3. On the <**xyz page**>, define the following configuration values:
      - Auto-discovery URL: `https://authentik.company/application/o/xenorchestra/.well-known/openid-configuration`
      - Client indentifier (key): the Client ID from your notes
      - Client secret: the Client Secret from your notes
      - etc etc

Author: Thats a great idea. I will change that 👍 

There's a screenshot of the configuration page in Xen Orchestra. All settings can be accessed within the same page.
![Screenshot XOA](https://github.com/goauthentik/authentik/assets/60177408/e911ba86-dd59-4af1-bc1d-97dde2a8395f)

4. Set the `Client identifier (key)` to the Client ID from your notes.
5. Set the `Client secret` to the Client Secret from your notes.
6. Check the `Fill information (optional)`-Checkbox to open the Advanced menu.
7. Set the `Authorization URL` to `https://authentik.company/application/o/authorize/`
8. Set the `Callback URL` to `https://xenorchestra.company/signin/oidc/callback`
9. Set the `Issuer` to `https://authentik.company/application/o/xenorchestra/`
10. Set the `Token URL` to `https://authentik.company/application/o/token/`
11. Set the `User info URL` to `https://authentik.company/application/o/userinfo/`
12. Set the `Username field` to `username`
13. Set the `Scopes` to `openid profile email`
14. Enable the `auth-oidc`-Plugin by toggling the switch above the configuration.
15. You should be able to login with OIDC.

:::note

> You need to create the user with the according username in XenOrchestra, before you are able to login using OIDC.

Reviewer: ```suggestion
You need to create the user with the according username in Xen Orchestra, before you are able to login using OIDC.
```

Optionally apply access restrictions to the application.

## Xen Orchestra configuration

Xen Orchestra allows the configuration of the OpenID Connect authentication in the plugin-section.
All of the URLs mentioned below can be copied & pasted from authentik (_Applications -> Providers -> *the provider created earlier*_).

1. Navigate to Settings -> Plugins
2. Scroll to **auth-oidc** and click on the **+** icon on the right hand side.
3. Configure the auth-oidc plugin with the following configuration values:

-   Set the `Auto-discovery URL` to `https://authentik.company/application/o/xenorchestra/.well-known/openid-configuration`.
-   Set the `Client identifier (key)` to the Client ID from your notes.
-   Set the `Client secret` to the Client Secret from your notes.
-   Check the `Fill information (optional)`-Checkbox to open the advanced menu.
-   Set the `Authorization URL` to `https://authentik.company/application/o/authorize/`
-   Set the `Callback URL` to `https://xenorchestra.company/signin/oidc/callback`
-   Set the `Issuer` to `https://authentik.company/application/o/xenorchestra/`
-   Set the `Token URL` to `https://authentik.company/application/o/token/`

> -   Set the `User info URL` to `https://authentik.company/application/o/userinfo/`

Reviewer 2: These URLs shouldn't have to be set manually as they can all be retrieved from the Auto-discovery URL above

Author: Oh I didn't notice that, when I was setting my Xen Orchestra plugin up.
I did some testing and can confirm, that those settings can be left blank in the Xen Orchestra configuration.

I've removed those lines now 👍

-   Set the `Username field` to `username`
-   Set the `Scopes` to `openid profile email`

4. Enable the `auth-oidc`-Plugin by toggling the switch above the configuration.
5. You should be able to login with OIDC.

> :::note

Reviewer 2: I think this should probably be moved up to make it less likely for people to miss it

Author: Unfortunately Xen Orchestra displays an error message immediately, if the plugin is enabled before setting the config values

<img width="299" alt="Screenshot of the error message when enabling the plugin first" src="https://github.com/goauthentik/authentik/assets/60177408/15545acf-3d1f-4ece-9ad3-581cb8bb4c5a">

Reviewer 2: ah I just saw the change you made to the note, disregard this
